### PR TITLE
Support for market orders in Waves.exchange

### PR DESCRIPTION
**Pull request by peterwilli**
# Intro

The waves.exchange docs claims [there's no support for market orders](https://docs.waves.exchange/en/ccxt/#createorder) on the Waves CCXT API.

Screenshot_20220130_223659

However, they didn't claim it's not supported in their own API. I needed this feature for my trading bot, and I decided to give it a go based on my previous assumption it could still be there.

I launched Waves.exchange in testnet and I managed to see how the market order worked and replicated it in this PR.

Some notes:

1. Unlike any other exchange I encountered, Waves forces you to set a price for market orders (used as minimum or maximum depending on if you sell or buy). Based on tests, I found out Waves.exchange (the web front-end) looks for prices in order book and bases the price that goes with the market order on the state of the order book. I did something similar, where I take the 4th cheapest/most expensive (again, depending on buy/sell) from the order book. 

    I took the 4th to give some leverage when bandwidth/slow internet will make an order arrive at Waves when the best option already left the order book (this rejects your market order). Note that Waves still finds the best price, so you won't have a disadvantage when you set a slightly worse price from the order book.

2. Waves.exchange claims in their docs market orders aren't supported. If this gets merged we probably need to inform them.

### CODE REVIEW

**Comment by kroitor** [comment]
Unfortunately, we don't really emulate the above functionality by checking the orderbook – that kind of emulation is out of the scope of the library and should be done in the userland. We should remove the call to the orderbook and make the user provide the price by requiring the `price` argument even for market orders.

For example:

- https://github.com/ccxt/ccxt/wiki/Manual#market-buys
- https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L3366

In other words, we should let the user do the orderbook call in the userland, and then expect a `price` argument in `createOrder` for market orders. If the user does not provide it, we should throw a meaningful exception as shown in Huobi and in many other places around the lib. Let me know if you can change this PR taking the above into account, or if you want my assistance on that. Thanks again!

**Comment by peterwilli** [comment]
> Unfortunately, we don't really emulate the above functionality by checking the orderbook – that kind of emulation is out of the scope of the library and should be done in the userland. We should remove the call to the orderbook and make the user provide the price by requiring the `price` argument even for market orders.
> 
> For example:
> 
>     * https://github.com/ccxt/ccxt/wiki/Manual#market-buys
> 
>     * https://github.com/ccxt/ccxt/blob/master/js/huobi.js#L3366
> 
> 
> In other words, we should let the user do the orderbook call in the userland, and then expect a `price` argument in `createOrder` for market orders. If the user does not provide it for market-type order, then we should throw a meaningful exception as shown in Huobi and in many other places around the lib. Let me know if you can change this PR taking the above into account, or if you want my assistance on that. Thanks again!

Yeah I was thinking of something like that as well. I guess that makes sense. It will slightly complicate things for me as I have to make a PR for freqtrade (the userland) as well, but what you say makes sense and I'll adjust it.

**Comment by peterwilli** [comment]
Would requiring price as `param` work? I don't think I can use the price argument when calling something like `create_market_sell_order`

**Comment by kroitor** [comment]
> Would requiring price as `param` work? I don't think I can use the price argument when calling something like `create_market_sell_order`

You can't use `create_order` there? Despite that `create_market_sell_order` does not have ` price` argument, you can change `create_market_sell_order(symbol, amount)` cause it is pretty much just a shorthand for `create_order(symbol, 'market', 'sell', amount, price)`. Let me know if this doesn't work:

```Python
response = await exchange.create_order(symbol, 'market', 'sell', amount, 10)
```

**Comment by peterwilli** [comment]
> response = await exchange.create_order(symbol, 'market', 'sell', amount, 10)

Thanks, works fine. I didn't know that was a thing. Also, I now realized much more exchanges require price for market orders, I didn't know that was a thing! Thought it was some stubborn Waves standard

**Comment by peterwilli** [comment]
@kroitor I made a new commit that removes the automatic order book traversal and lets users provide their own price. If you don't do that you get a helpful error.
